### PR TITLE
feat(storage-controller): add node shards api

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

For control-plane managed tenants, we have the page in the admin console that lists all tenants on a specific pageserver. But for storage-controller managed ones, we don't have that functionality for now.

## Summary of changes

Adds an API that lists all shards on a given node (intention + observed)

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: control_plane/storcon_cli/src/main.rs [hunk 1]
```diff
@@ -77,6 +77,11 @@ enum Command {
         #[arg(long)]
         scheduling: Option<ShardSchedulingPolicyArg>,
     },
+    /// Print details about a particular node, including all shards attached to it.
+    NodeAttached {
```
Comment (arpad-m) [inline control_plane/storcon_cli/src/main.rs:81]: I wonder if one could expose it as an optional `node` param of `Tenants` instead.

File: test_runner/fixtures/common_types.py [hunk 2]
```diff
@@ -140,6 +140,14 @@ def __str__(self) -> str:
         return self.id.hex()
 
 
+class NodeId(Id):
+    def __repr__(self) -> str:
+        return f'`TenantId("{self.id.hex()}")'
```
Comment (arpad-m) [inline test_runner/fixtures/common_types.py:145]:
```suggestion
        return f'`NodeId("{self.id.hex()}")'
```

File: control_plane/storcon_cli/src/main.rs [hunk 3]
```diff
@@ -508,6 +513,29 @@ async fn main() -> anyhow::Result<()> {
...
+            let describe_response = storcon_client
+                .dispatch::<(), NodeAttachedResponse>(
+                    Method::GET,
+                    format!("control/v1/node/{node}/attached"),
+                    None,
+                )
+                .await?;
+            let shards = describe_response.shards;
+            let mut table = comfy_table::Table::new();
+            table.set_header(["Shard", "Primary/Secondary"]);
```
Comment (arpad-m) [inline control_plane/storcon_cli/src/main.rs:526]: I wonder if we could have a third column here for whether the attached state has been observed or is just intended by the storage controller. Or maybe that's not useful, idk. What do you think?

**Review by arpad-m — DISMISSED** [review]

**Comment by jcsp** [comment]
Please fill out the PR template:  the "Problem" part is important too as it tells us why this change is needed.

File: test_runner/fixtures/neon_fixtures.py [hunk 4]
```diff
@@ -2553,6 +2553,30 @@ def tenant_describe(self, tenant_id: TenantId):
...
+        """
+        response = self.request(
+            "GET",
+            f"{self.api}/control/v1/node",
+            headers=self.headers(TokenScope.ADMIN),
+        )
+        response.raise_for_status()
+        return response.json()
+
+    def node_attached(self, node_id: NodeId):
```
Comment (jcsp) [inline test_runner/fixtures/neon_fixtures.py:2568]: "attached" means nodes which are fully attached (excluding secondaries).  Let's tweak this name to something like "node_shards" if it should include all secondaries as well.

**Comment by skyzh** [comment]
Now the API is called node_shards, and returns both observed + intended states.

File: libs/pageserver_api/src/controller_api.rs [hunk 5]
```diff
@@ -101,6 +101,19 @@ pub struct TenantDescribeResponse {
...
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NodeShardResponse {
+    pub node_id: NodeId,
+    pub shards: Vec<NodeShard>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NodeShard {
+    pub tenant_shard_id: TenantShardId,
+    pub is_observed_secondary: Option<bool>,
```
Comment (jcsp) [inline libs/pageserver_api/src/controller_api.rs:113]: Why `Option<bool>` rather than just bool?

**Review by jcsp — APPROVED** [review]

File: libs/pageserver_api/src/controller_api.rs (hunk 5 above)
Comment (skyzh) [reply libs/pageserver_api/src/controller_api.rs:113]: None == the tenant doesn't exist on this node. I have a commit but it didn't get pushed...?

Comment (skyzh) [reply libs/pageserver_api/src/controller_api.rs:113]: None == the tenant doesn't exist on this node. I have a commit to add the comments but it didn't get pushed...?

Comment (jcsp) [reply libs/pageserver_api/src/controller_api.rs:113]: Ah, so we're saying Some(false) means "this shard is attached", but None means "this shard does not appear in the intent state" (but it must appear in the observed state or we wouldn't have constructed a NodeShard).

This is a bit subtle, but okay - the CLI output is still easy to understand.

Comment (skyzh) [reply libs/pageserver_api/src/controller_api.rs:114]: I couldn't find a better naming for that field :( anyways, the updated comments:
```
    /// Whether the shard is observed secondary on a specific node. True = yes, False = no, None = not on this node.
    pub is_observed_secondary: Option<bool>,
```

**Pull request merged**